Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+#include "Firestore/core/src/firebase/firestore/remote/remote_store.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/delayed_constructor.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+using firestore::Error;
+using api::DocumentReference;
+using api::DocumentSnapshot;
+using api::QuerySnapshot;
+using api::Settings;
+using api::SnapshotMetadata;
+using api::ThrowIllegalState;
+using auth::CredentialsProvider;
+using auth::User;
+using core::DatabaseInfo;
+using core::ListenOptions;
+using core::EventManager;
+using core::Query;
+using core::QueryListener;
+using core::ViewSnapshot;
+using local::LruParams;
+using model::DatabaseId;
+using model::Document;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::MaybeDocument;
+using model::Mutation;
+using model::OnlineState;
+using remote::Datastore;
+using remote::RemoteStore;
+using util::Path;
+using util::AsyncQueue;
+using util::DelayedConstructor;
+using util::DelayedOperation;
+using util::Executor;
+using util::Status;
+using util::StatusCallback;
+using util::StatusOr;
+using util::StatusOrCallback;
+using util::TimerId;
+
+std::shared_ptr<FirestoreClient> FirestoreClient::Create(
+    const DatabaseInfo& database_info,
+    const api::Settings& settings,
+    std::shared_ptr<auth::CredentialsProvider> credentials_provider,
+    std::shared_ptr<util::Executor> user_executor,
+    std::shared_ptr<util::AsyncQueue> worker_queue) {
+  // Have to use `new` because `make_shared` cannot access private constructor.
+  std::shared_ptr<FirestoreClient> client(
+      new FirestoreClient(database_info, std::move(credentials_provider),
+                          std::move(user_executor), std::move(worker_queue)));
+  client->Initialize(settings);
+  return client;
+}
+
+FirestoreClient::FirestoreClient(
+    const DatabaseInfo& database_info,
+    std::shared_ptr<auth::CredentialsProvider> credentials_provider,
+    std::shared_ptr<util::Executor> user_executor,
+    std::shared_ptr<util::AsyncQueue> worker_queue)
+    : database_info_(database_info),
+      credentials_provider_(std::move(credentials_provider)),
+      worker_queue_(std::move(worker_queue)),
+      user_executor_(std::move(user_executor)) {
+}
+
+void FirestoreClient::Initialize(const api::Settings& settings) {
+  if (client_initialized_) {
+    return;
+  }
+  auto user_promise = std::make_shared<std::promise<User>>();
+  bool credentials_initialized = false;
+
+  std::weak_ptr<FirestoreClient> weak_self(shared_from_this());
```

Comment:
"self" is an Objective-C-ism. Prefer "this" names like weak_self and shared_self.

Reply:
Done.